src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Authentication.Certificate
+{
+    internal class CertificateAuthenticationHandler : AuthenticationHandler<CertificateAuthenticationOptions>
+    {
+        private static readonly Oid ClientCertificateOid = new Oid("1.3.6.1.5.5.7.3.2");
+
+        public CertificateAuthenticationHandler(
+            IOptionsMonitor<CertificateAuthenticationOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock) : base(options, logger, encoder, clock)
+        {
+        }
+
+        /// <summary>
+        /// The handler calls methods on the events which give the application control at certain points where processing is occurring.
+        /// If it is not provided a default instance is supplied which does nothing when the methods are called.
+        /// </summary>
+        protected new CertificateAuthenticationEvents Events
+        {
+            get { return (CertificateAuthenticationEvents)base.Events; }
+            set { base.Events = value; }
+        }
+
+        /// <summary>
+        /// Creates a new instance of the events instance.
+        /// </summary>
+        /// <returns>A new instance of the events instance.</returns>
+        protected override Task<object> CreateEventsAsync() => Task.FromResult<object>(new CertificateAuthenticationEvents());
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            // You only get client certificates over HTTPS
+            if (!Context.Request.IsHttps)
+            {
+                return AuthenticateResult.NoResult();
+            }
+
+            var clientCertificate = await Context.Connection.GetClientCertificateAsync();
+
+            // This should never be the case, as cert authentication happens long before ASP.NET kicks in.
+            if (clientCertificate == null)
+            {
+                Logger.LogDebug("No client certificate found.");
+                return AuthenticateResult.NoResult();
+            }
+
+            // If we have a self signed cert, and they're not allowed, exit early and not bother with
+            // any other validations.
+            if (clientCertificate.IsSelfSigned() &&
+                !Options.AllowedCertificateTypes.HasFlag(CertificateTypes.SelfSigned))
+            {
+                Logger.LogWarning("Self signed certificate rejected, subject was {0}", clientCertificate.Subject);
```

Comment:
Needs proper structured logging instead of ordinal parameters.

Reply:
Structured logging added